crates/ruma-state-res/tests/it/main.rs

```diff
@@ -0,0 +1,249 @@
+};
+use ruma_events::{StateEventType, TimelineEventType};
+use ruma_state_res::{resolve, Event, StateMap};
+use serde::{Deserialize, Serialize};
+use serde_json::{to_value as to_json_value, value::RawValue as RawJsonValue, Value as JsonValue};
+use tracing_subscriber::{util::SubscriberInitExt, EnvFilter};
+
+/// A persistent data unit
+#[derive(Deserialize, Clone)]
+struct Pdu {
+    event_id: OwnedEventId,
+    room_id: OwnedRoomId,
+    sender: OwnedUserId,
+    origin_server_ts: MilliSecondsSinceUnixEpoch,
+    #[serde(rename = "type")]
+    kind: TimelineEventType,
+    content: Box<RawJsonValue>,
+    state_key: Option<String>,
+    prev_events: Vec<OwnedEventId>,
+    auth_events: Vec<OwnedEventId>,
+    redacts: Option<OwnedEventId>,
+}
+
+impl Event for Pdu {
+    type Id = OwnedEventId;
+
+    fn event_id(&self) -> &Self::Id {
+        &self.event_id
+    }
+
+    fn room_id(&self) -> &RoomId {
+        &self.room_id
+    }
+
+    fn sender(&self) -> &UserId {
+        &self.sender
+    }
+
+    fn origin_server_ts(&self) -> MilliSecondsSinceUnixEpoch {
+        self.origin_server_ts
+    }
+
+    fn event_type(&self) -> &TimelineEventType {
+        &self.kind
+    }
+
+    fn content(&self) -> &RawJsonValue {
+        &self.content
+    }
+
+    fn state_key(&self) -> Option<&str> {
+        self.state_key.as_deref()
+    }
+
+    fn prev_events(&self) -> Box<dyn DoubleEndedIterator<Item = &Self::Id> + '_> {
+        Box::new(self.prev_events.iter())
+    }
+
+    fn auth_events(&self) -> Box<dyn DoubleEndedIterator<Item = &Self::Id> + '_> {
+        Box::new(self.auth_events.iter())
+    }
+
+    fn redacts(&self) -> Option<&Self::Id> {
+        self.redacts.as_ref()
+    }
+}
+
+/// Extract `.content.room_version` from a PDU
+#[derive(Deserialize)]
+struct ExtractRoomVersion {
+    room_version: RoomVersionId,
+}
+
+/// Type describing a resolved state event
+#[derive(Serialize)]
+struct ResolvedState {
+    kind: StateEventType,
+    state_key: String,
+    event_id: OwnedEventId,
+    content: JsonValue,
+}
+
+impl PartialEq for ResolvedState {
+    fn eq(&self, other: &Self) -> bool {
+        self.kind == other.kind
+            && self.state_key == other.state_key
+            && self.event_id == other.event_id
+    }
+}
+
+impl Eq for ResolvedState {}
+
+impl Ord for ResolvedState {
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ordering::Equal
+            .then(self.kind.cmp(&other.kind))
+            .then(self.state_key.cmp(&other.state_key))
+            .then(self.event_id.cmp(&other.event_id))
+    }
+}
+
+impl PartialOrd for ResolvedState {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Results of a test
+struct TestResults {
+    /// The resolved state of the room
+    resolved_state: BTreeSet<ResolvedState>,
+}
+
+/// Test a JSON file containing a list of PDUs and return the results
+fn run<P>(path: P) -> Result<TestResults, Box<dyn Error>>
```

Comment:
Do you have any suggestions?

Reply:
Something simple like `run_snapshot_test` maybe?

Also I am wondering if there is any value in using generics here. It is a private function and we will likely always use `&str`.

Reply:
Ah no there is no snapshot in this. `resolve_files` then?

Reply:
> Also I am wondering if there is any value in using generics here.

Habit. I can get rid of them if you want.

---

I kinda chose a vague name on purpose because this function is basically like a `main` function for the snapshot tests; it does everything, so naming it something specific doesn't feel right. It doesn't just do stateres, it also loads data, it also can do stateres incrementally, it also restructures the data to be easier to read in snapshot tests, and in the future it may do more things like generate graphviz graphs, assert on failed events, etc. Fwiw there's also a function in the pre-existing tests that serves a similar purpose and it's just called `do_check` which has similar name vibes IMO.

Reply:
That's the thing, there is no reason to have a single function that does everything, this is not a CLI. If you want to do different things, implement different functions and refactor the common code in separate functions.

Currently all that it does is perform state res on a bunch of files, and it should probably only ever do that. If the sole purpose of changing the format of the output is to make it look pretty in the snapshot, it should probably be a separate function that is called in `snapshot_test!`.

And that `do_check` function is problematic too, I have no idea what it does at first glance.

Reply:
>> Also I am wondering if there is any value in using generics here.
>
> Habit. I can get rid of them if you want.

Please do, they just make the functions harder to read.

Reply:
> If the sole purpose of changing the format of the output is to make it look pretty in the snapshot, it should probably be a separate function that is called in `snapshot_test!`.

I am deliberately keeping as much code out of the macro body as possible because language server features don't work very well inside them. I don't think it's that unreasonable to have a regular function that does all the work to generate snapshot test data, and for the macro to only contain assertions on its return value. This isn't a publicly exposed library function that needs to be maximally flexible or anything.

Reply:
In any case I still believe that `resolve_state_in_files` is a clearer name for the function than `run`, and it is the general purpose of the function, even if it transforms the data.

Reply:
I've gone with `test_resolve`.